### DISCUSSION TITLE
mpremote oddity

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#15091] [General]
I have an application which opens a file in "rb" mode. The file, a `pgm` graphics file, contains four lines of `\n` terminated ASCII text followed by binary data. This works if the file and the Python code are located on the RP2 host. If they are located on the PC and run via `mpremote mount .` it fails.

I have produced the following repro. It works on CPython, on the Unix build, and on RP2 if located on the RP2 Flash.
```py
import os

fn = "delete_me"
with open(fn, "wb") as f:
    f.write(b"hello\n\xde\xad\xbe\xef")
with open(fn, "rb") as f:
    print(f.readline())  # Odd behaviour here
    print(f.read(4))
os.unlink(fn)
```
Running with the application on the RP2 it behaves as expected:
```py
>>> import rats15
b'hello\n'
b'\xde\xad\xbe\xef'
>>> 
```
Running on RP2 with the source on the PC:
```py
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "rats15.py", line 7, in <module>
  File "<stdin>", line 147, in readline
TypeError: unsupported types for __add__: 'str', 'bytes'
```
Note that the file is written correctly, the error occurs on read. Further, it is the `readline` which causes the problem. If I write a function to do the same job, reading and discarding bytes until `\n`, the fault disappears.

I'm not sure if this is an `rpremote` bug or a known limitation?

**Comment by cwalther** [comment]
At a cursory glance, that looks like a bug on this line to me:
https://github.com/micropython/micropython/blob/cfd5a8ea3a80930a79dbae6a6aff54587393d0b8/tools/mpremote/mpremote/transport_serial.py#L800

That should probably be `l = '' if self.is_text else b''` (and the check for `\n` below that needs a variant too).

**Comment by scruss** [comment]
> I'm not sure if this is an `mpremote` bug or a known limitation?

Couldn't reproduce this behaviour with mpremote 1.22.0 and MicroPython v1.22.2 on both Raspberry Pi Pico and STM F4 Discovery Board. `rats15.py` executed as expected when stored locally (`mpremote a1 run rats15.py`) or called from the REPL (`import rats15`).

**Reply by peterhinch** [reply]
I evidently didn't properly explain my test method. At the PC I issue
```bash
$ mpremote mount .
```
At the REPL:
```py
MicroPython v1.22.2 on 2024-02-22; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>> import rats15
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "rats15.py", line 7, in <module>
  File "<stdin>", line 147, in readline
TypeError: unsupported types for __add__: 'str', 'bytes'
>>> 
```
I am running mpremote 1.22.0. I get the same outcome on a Pyboard 1.1.

It's not clear to me which file has line 147...

**Comment by peterhinch** [comment]
I have raised https://github.com/micropython/micropython/issues/15098

**Comment by cwalther** [comment]
> It's not clear to me which file has line 147...

I’m pretty sure it’s the string literal (`fs_hook_code`) that my link pointed into, after the compression done on line 950 of _transport_serial.py_.

**Reply by peterhinch** [reply]
I think that the type returned by the call to `.read()` depends on whether the file was opened "rb" or "r" so it's not obvious (to me) what is the best way to fix this.
